# Semana 6 

## 0. carga de datos 

In [1]:
import pandas as pd

df = pd.read_csv("../../data\S04_PD_Lozano_DatasetImputacionInicial.csv")

print(df.shape)
df.head()

<>:3: SyntaxWarning: invalid escape sequence '\S'
<>:3: SyntaxWarning: invalid escape sequence '\S'
C:\Users\diego\AppData\Local\Temp\ipykernel_23920\2539300773.py:3: SyntaxWarning: invalid escape sequence '\S'
  df = pd.read_csv("../../data\S04_PD_Lozano_DatasetImputacionInicial.csv")


(589386, 22)


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,...,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente,correo_cliente_faltante,comentario_cliente_faltante
0,P260259,C129774,2/3/2026,Marketplace,Transferencia,VILLAVICENCIO,5000102.0,electronica,Teclado mecánico,165200.0,...,59.0,cliente160259@yahoo.es,bajo,Oro,Pésima experiencia con el empaque,8.3,2026-03-04,0.0,0.0,0.0
1,"P700975,C159363,11/10/2025,Web, Transferencia,...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,P641357,C118866,31/7/2025,App,Efectivo,ibague,7300103.0,Hogar,Olla a presión,756600.0,...,57.0,cliente541357@yahoo.es,medio,Oro,Muy buen precio,13.8,2026-07-12,1.0,0.0,0.0
3,"P288663,C101979,2025-09-09,Marketplace,Transfe...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,P296684,C147052,21/12/2025,Tienda,Transferencia,Manizales,1700110.0,Hogar,Lámpara de escritorio,749700.0,...,68.0,NaN,medio,Oro,NaN,13.9,2026-06-10,0.0,1.0,1.0


## 1. Correcion de datos

In [2]:
for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]:
    print(col, sorted(df[col].dropna().unique()))

ciudad_tienda [' Barranquilla', ' Bogotá', ' Bucaramanga', ' Cali', ' Cartagena', ' Cúcuta', ' Ibagué', ' Manizales', ' MedellÃ\xadn', ' Pereira', ' Santa Marta', ' Villavicencio', 'B/manga', 'B/quilla', 'BAQ', 'BARRANQUILLA', 'BGA', 'BOG', 'BOGOTA', 'BOGOTA DC', 'BOGOTÃ\x81', 'BUCARAMANGA', 'Barranquilla', 'Barranquilla ', 'Bogota', 'Bogota D.C.', 'Bogotá', 'Bogotá ', 'Bucaramanga', 'Bucaramanga ', 'CALI', 'CALI ', 'CARTAGENA', 'CTG', 'CUCUTA', 'Cali', 'Cali ', 'Cartagena', 'Cartagena ', 'Cartagena de Indias', 'Cucuta', 'Cúcuta', 'Cúcuta ', 'CÚCUTA', 'IBAGUE', 'IBAGUÉ', 'Ibague', 'Ibagué', 'Ibagué ', 'MANIZALES', 'MDE', 'MEDELLIN', 'MEDELLÃ\x8dN', 'MZL', 'Manizales', 'Manizales ', 'Medellin', 'MedellÃ\xadn', 'MedellÃ\xadn ', 'PEREIRA', 'Pereira', 'Pereira ', 'SANTA MARTA', 'SMR', 'Santa Marta', 'Santa Marta ', 'SantaMarta', 'Santiago de Cali', 'Sta. Marta', 'VILLAVICENCIO', 'Villavicencio', 'Villavicencio ', 'barranquilla', 'bogota', 'bogota d.c.', 'bucaramanga', 'cali', 'c

### Corrección de encoding en columnas categóricas

Al inspeccionar los valores únicos de `ciudad_tienda`, `categoria_producto`, `canal_compra` y `metodo_pago`, se detectaron caracteres corruptos en las tildes (ej. `Bogotá`, `Electrónica`), un patrón conocido como *mojibake*. Esto ocurre cuando el archivo fue codificado en UTF-8 pero leído/guardado con una codificación distinta (Windows-1252).

Se aplicó `str.encode('cp1252', errors='ignore').str.decode('utf-8', errors='ignore')` a las 4 columnas para revertir la doble codificación y recuperar los caracteres originales (tildes y ñ). Se usó `errors='ignore'` para no interrumpir el proceso ante bytes atípicos, que se revisan por separado.

In [3]:
cols_texto = ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]

for col in cols_texto:
    df[col] = df[col].str.encode('cp1252', errors='ignore').str.decode('utf-8', errors='ignore')

# Verifica que se arregló
for col in cols_texto:
    print(col, sorted(df[col].dropna().unique()))

ciudad_tienda [' Barranquilla', ' Bogotá', ' Bucaramanga', ' Cali', ' Cartagena', ' Cúcuta', ' Ibagué', ' Manizales', ' Medellín', ' Pereira', ' Santa Marta', ' Villavicencio', 'B/manga', 'B/quilla', 'BAQ', 'BARRANQUILLA', 'BGA', 'BOG', 'BOGOT', 'BOGOTA', 'BOGOTA DC', 'BUCARAMANGA', 'Barranquilla', 'Barranquilla ', 'Bogota', 'Bogota D.C.', 'Bogotá', 'Bogotá ', 'Bucaramanga', 'Bucaramanga ', 'CALI', 'CALI ', 'CARTAGENA', 'CTG', 'CUCUTA', 'Cali', 'Cali ', 'Cartagena', 'Cartagena ', 'Cartagena de Indias', 'Cucuta', 'CÚCUTA', 'Cúcuta', 'Cúcuta ', 'IBAGUE', 'IBAGUÉ', 'Ibague', 'Ibagué', 'Ibagué ', 'MANIZALES', 'MDE', 'MEDELLIN', 'MEDELLN', 'MZL', 'Manizales', 'Manizales ', 'Medellin', 'Medellín', 'Medellín ', 'PEREIRA', 'Pereira', 'Pereira ', 'SANTA MARTA', 'SMR', 'Santa Marta', 'Santa Marta ', 'SantaMarta', 'Santiago de Cali', 'Sta. Marta', 'VILLAVICENCIO', 'Villavicencio', 'Villavicencio ', 'barranquilla', 'bogota', 'bogota d.c.', 'bucaramanga', 'cali', 'cartagena', 'cucuta', 'ibague', 'm

### Catálogo de variantes a agrupar (por columna)

#### ciudad_tienda (12 valores reales)
- **Bogotá**: `Bogotá`, `Bogotá `, ` Bogotá`, `BOGOTA`, `Bogota`, `bogota`, `BOGOTA DC`, `Bogota D.C.`, `bogota d.c.`, `BOG`, `BOGOT` (corrupto)
- **Medellín**: `Medellín`, `Medellín `, `MEDELLIN`, `Medellin`, `medellin`, `MDE`, `MEDELLN` (corrupto)
- **Cali**: `Cali`, `Cali `, ` Cali`, `CALI`, `CALI `, `cali`, `Santiago de Cali`
- **Cartagena**: `Cartagena`, `Cartagena `, ` Cartagena`, `CARTAGENA`, `cartagena`, `Cartagena de Indias`
- **Barranquilla**: `Barranquilla`, `Barranquilla `, ` Barranquilla`, `BARRANQUILLA`, `barranquilla`, `B/quilla`, `BAQ`
- **Bucaramanga**: `Bucaramanga`, `Bucaramanga `, ` Bucaramanga`, `BUCARAMANGA`, `bucaramanga`, `B/manga`, `BGA`
- **Cúcuta**: `Cúcuta`, `Cúcuta `, ` Cúcuta`, `CÚCUTA`, `CUCUTA`, `Cucuta`, `cucuta`
- **Ibagué**: `Ibagué`, `Ibagué `, ` Ibagué`, `IBAGUÉ`, `IBAGUE`, `Ibague`, `ibague`
- **Manizales**: `Manizales`, `Manizales `, ` Manizales`, `MANIZALES`, `manizales`, `MZL`
- **Pereira**: `Pereira`, `Pereira `, ` Pereira`, `PEREIRA`, `pereira`
- **Santa Marta**: `Santa Marta`, `Santa Marta `, ` Santa Marta`, `SANTA MARTA`, `santa marta`, `SantaMarta`, `santamarta`, `Sta. Marta`, `SMR`
- **Villavicencio**: `Villavicencio`, `Villavicencio `, ` Villavicencio`, `VILLAVICENCIO`, `villavicencio`

#### categoria_producto (6 valores reales)
- **Belleza**: `Belleza`, `Belleza `, ` Belleza`, `BELLEZA`, `belleza`
- **Deportes**: `Deportes`, `Deportes `, ` Deportes`, `DEPORTES`, `deportes`
- **Electrónica**: `Electrónica`, `Electrónica `, ` Electrónica`, `ELECTRÓNICA`, `electrónica`, `electronica`
- **Hogar**: `Hogar`, `Hogar `, ` Hogar`, `HOGAR`, `hogar`
- **Juguetería**: `Juguetería`, `Juguetería `, ` Juguetería`, `JUGUETERÍA`, `jugueteria`, `juguetería`, `JUGUETERA` (corrupto)
- **Moda**: `Moda`, `Moda `, ` Moda`, `MODA`, `moda`

#### canal_compra (4 valores reales)
- **App**: `App`, `App `, ` App`, `APP`, `app`
- **Marketplace**: `Marketplace`, `Marketplace `, ` Marketplace`, `MARKETPLACE`, `marketplace`
- **Tienda**: `Tienda`, `Tienda `, ` Tienda`, `TIENDA`, `tienda`
- **Web**: `Web`, `Web `, ` Web`, `WEB`, `web`

#### metodo_pago (4 valores reales)
- **Efectivo**: `Efectivo`, `Efectivo `, ` Efectivo`, `EFECTIVO`, `efectivo`
- **PayPal**: `PayPal`, `PayPal `, ` PayPal`, `PAYPAL`, `paypal`
- **Tarjeta**: `Tarjeta`, `Tarjeta `, ` Tarjeta`, `TARJETA`, `tarjeta`
- **Transferencia**: `Transferencia`, `Transferencia `, ` Transferencia`, `TRANSFERENCIA`, `transferencia`

**Nota**: `canal_compra` y `metodo_pago` se resuelven completamente con la normalización estándar (`.str.strip().str.lower().str.normalize('NFKD')...`). `ciudad_tienda` y `categoria_producto` requieren, además, un diccionario de mapeo manual para abreviaturas, alias y los 3 casos de encoding corrupto.

In [4]:
df["ciudad_tienda"] = df["ciudad_tienda"].str.strip().str.title()
df["categoria_producto"] = df["categoria_producto"].str.strip().str.title()
df["canal_compra"] = df["canal_compra"].str.strip().str.title()
df["metodo_pago"] = df["metodo_pago"].str.strip().str.title()

# Confirma que no se perdieron filas
print(df.shape[0])

589386


In [5]:
for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]:
    print(col, sorted(df[col].dropna().unique()))

ciudad_tienda ['B/Manga', 'B/Quilla', 'Baq', 'Barranquilla', 'Bga', 'Bog', 'Bogot', 'Bogota', 'Bogota D.C.', 'Bogota Dc', 'Bogotá', 'Bucaramanga', 'Cali', 'Cartagena', 'Cartagena De Indias', 'Ctg', 'Cucuta', 'Cúcuta', 'Ibague', 'Ibagué', 'Manizales', 'Mde', 'Medellin', 'Medelln', 'Medellín', 'Mzl', 'Pereira', 'Santa Marta', 'Santamarta', 'Santiago De Cali', 'Smr', 'Sta. Marta', 'Villavicencio']
categoria_producto ['Belleza', 'Deportes', 'Electronica', 'Electrónica', 'Hogar', 'Juguetera', 'Jugueteria', 'Juguetería', 'Moda']
canal_compra ['App', 'Marketplace', 'Tienda', 'Web']
metodo_pago ['Efectivo', 'Paypal', 'Tarjeta', 'Transferencia']


### Diccionarios de corrección

In [6]:
correcciones_ciudad = {
    "Bog": "Bogotá",
    "Bogot": "Bogotá",           # caso corrupto (encoding)
    "Bogota": "Bogotá",
    "Bogota D.C.": "Bogotá",
    "Bogota Dc": "Bogotá",
    "Mde": "Medellín",
    "Medellin": "Medellín",
    "Medelln": "Medellín",       # caso corrupto (encoding)
    "Ctg": "Cartagena",
    "Cartagena De Indias": "Cartagena",
    "Cucuta": "Cúcuta",
    "Baq": "Barranquilla",
    "B/Quilla": "Barranquilla",
    "Bga": "Bucaramanga",
    "B/Manga": "Bucaramanga",
    "Ibague": "Ibagué",
    "Mzl": "Manizales",
    "Santamarta": "Santa Marta",
    "Sta. Marta": "Santa Marta",
    "Smr": "Santa Marta",
    "Santiago De Cali": "Cali",
}

correcciones_categoria = {
    "Electronica": "Electrónica",
    "Jugueteria": "Juguetería",
    "Juguetera": "Juguetería",   # caso corrupto (encoding)
}

df["ciudad_tienda"] = df["ciudad_tienda"].replace(correcciones_ciudad)
df["categoria_producto"] = df["categoria_producto"].replace(correcciones_categoria)

for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]:
    print(col, sorted(df[col].dropna().unique()))

ciudad_tienda ['Barranquilla', 'Bogotá', 'Bucaramanga', 'Cali', 'Cartagena', 'Cúcuta', 'Ibagué', 'Manizales', 'Medellín', 'Pereira', 'Santa Marta', 'Villavicencio']
categoria_producto ['Belleza', 'Deportes', 'Electrónica', 'Hogar', 'Juguetería', 'Moda']
canal_compra ['App', 'Marketplace', 'Tienda', 'Web']
metodo_pago ['Efectivo', 'Paypal', 'Tarjeta', 'Transferencia']


## 2. nulos 


In [7]:
nulos_antes = df["fecha_pedido"].isnull().sum()
print("Nulos antes:", nulos_antes)

Nulos antes: 89706


In [8]:
meses_es_a_num = {
    "ene": "01", "feb": "02", "mar": "03", "abr": "04",
    "may": "05", "jun": "06", "jul": "07", "ago": "08",
    "sep": "09", "oct": "10", "nov": "11", "dic": "12"
}

In [12]:
for mes_es, mes_num in meses_es_a_num.items():
    df["fecha_pedido"] = df["fecha_pedido"].str.replace(f"-{mes_es}-", f"-{mes_num}-", regex=False)

In [13]:
prueba = pd.to_datetime(df["fecha_pedido"], dayfirst=True, errors="coerce")
mask_nuevos_nulos = prueba.isnull() & df["fecha_pedido"].notnull()

print(mask_nuevos_nulos.sum())
print(df.loc[mask_nuevos_nulos, "fecha_pedido"].unique()[:30])

173927
<StringArray>
['2026-04-18', '2026-13-05', '2025-05-03', '2025-05-17', '2025-04-08',
 '2025-07-24', '2026-02-27', '2025-09-17', '2025-02-18', '2025-01-31',
 '2025-10-01', '2025-01-19', '2026-04-21', '2025-09-13', '2026-03-18',
 '2025-03-08', '30/02/2025', '2025-01-15', '2025-04-15', '2025-11-15',
 '2025-09-24', '2026-01-20', '2025-12-16', '2025-02-12', '2025-03-19',
 '2025-05-30', '2025-01-05', '2025-04-28', '2026-01-09', '2025-10-08']
Length: 30, dtype: str


In [14]:
patron_iso = df["fecha_pedido"].str.match(r'^\d{4}-\d{2}-\d{2}$', na=False)
patron_slash = df["fecha_pedido"].str.match(r'^\d{1,2}/\d{1,2}/\d{4}$', na=False)
otros = df["fecha_pedido"].notna() & ~patron_iso & ~patron_slash

print("Formato ISO (YYYY-MM-DD):", patron_iso.sum())
print("Formato D/M/YYYY:", patron_slash.sum())
print("Otros / no clasificados:", otros.sum())
print(df.loc[otros, "fecha_pedido"].unique()[:30])

Formato ISO (YYYY-MM-DD): 160485
Formato D/M/YYYY: 339195
Otros / no clasificados: 0
<StringArray>
[]
Length: 0, dtype: str


In [15]:
fecha_final = pd.Series(pd.NaT, index=df.index, dtype='datetime64[ns]')

# Formato ISO: YYYY-MM-DD
fecha_final[patron_iso] = pd.to_datetime(df.loc[patron_iso, "fecha_pedido"], format="%Y-%m-%d", errors="coerce")

# Formato D/M/YYYY
fecha_final[patron_slash] = pd.to_datetime(df.loc[patron_slash, "fecha_pedido"], format="%d/%m/%Y", errors="coerce")

df["fecha_pedido"] = fecha_final

print("Nulos antes:", nulos_antes)
print("Nulos después:", df["fecha_pedido"].isnull().sum())

Nulos antes: 89706
Nulos después: 107488


In [16]:
nuevos_nulos = df["fecha_pedido"].isnull().sum() - nulos_antes
print("Nuevos nulos por fechas inválidas:", nuevos_nulos)

# Volvamos a ver qué texto original causó estos NaT
# (usando las máscaras que ya armamos, antes de la conversión)
invalidas_iso = patron_iso & df["fecha_pedido_original"].notna() if "fecha_pedido_original" in df.columns else None

Nuevos nulos por fechas inválidas: 17782


In [19]:
df_temp = pd.read_csv("../../data\S04_PD_Lozano_DatasetImputacionInicial.csv", usecols=["fecha_pedido"])

for mes_es, mes_num in meses_es_a_num.items():
    df_temp["fecha_pedido"] = df_temp["fecha_pedido"].str.replace(f"-{mes_es}-", f"-{mes_num}-", regex=False)

mask_ahora_nulas = df["fecha_pedido"].isnull() & df_temp["fecha_pedido"].notna()
print(mask_ahora_nulas.sum())
print(df_temp.loc[mask_ahora_nulas, "fecha_pedido"].value_counts().head(20))

<>:1: SyntaxWarning: invalid escape sequence '\S'
<>:1: SyntaxWarning: invalid escape sequence '\S'
C:\Users\diego\AppData\Local\Temp\ipykernel_23920\4142062038.py:1: SyntaxWarning: invalid escape sequence '\S'
  df_temp = pd.read_csv("../../data\S04_PD_Lozano_DatasetImputacionInicial.csv", usecols=["fecha_pedido"])


17782
fecha_pedido
00/00/0000    4557
31/04/2025    4516
30/02/2025    4369
2026-13-05    4340
Name: count, dtype: int64


### Hallazgo: fechas inválidas en fecha_pedido

Se identificaron 17,782 registros (≈3.02% del dataset) con fechas sintácticamente inválidas en `fecha_pedido`, agrupados en 4 patrones:
- `00/00/0000` (4,557 registros): fecha sentinela/placeholder.
- `31/04/2025` (4,516 registros): día inexistente para el mes (abril tiene 30 días).
- `30/02/2025` (4,369 registros): día inexistente para el mes (febrero no bisiesto).
- `2026-13-05` (4,340 registros): mes fuera de rango (no existe el mes 13).

Estas fechas se convirtieron a `NaT` mediante `errors="coerce"`, sumándose a los 89,706 nulos ya existentes desde la Semana 4, para un total de 107,488 valores nulos en la columna.

**Severidad: Alta** — el campo es crítico para análisis de tendencias temporales y reportes de oportunidad (dimensión Oportunidad), y el patrón sugiere un problema sistemático de origen (posible error de generación o placeholder mal usado), no ruido aleatorio.

## 3. Diagnostico duplicados 

In [28]:
df.duplicated().sum()

np.int64(18410)

### Duplicados exactos encontrados
Se encontraron 18,410 filas duplicadas exactas (idénticas en las 22 columnas), equivalentes a ≈3.12% del dataset.

In [29]:
df.duplicated(subset=["correo_cliente"], keep=False).sum()

np.int64(197944)

In [30]:
duplicados_correo = df[df["correo_cliente"].notna()].duplicated(subset=["correo_cliente"], keep=False).sum()
print(duplicados_correo)

26128


In [31]:
print(df["correo_cliente"].isna().sum())

171816


### Duplicados por correo encontrados

Al ejecutar `df.duplicated(subset=["correo_cliente"], keep=False).sum()` sin filtrar, se obtuvieron 197,944 registros. Sin embargo, este número está inflado porque pandas trata los valores `NaN` como iguales entre sí, y `correo_cliente` tiene una cantidad importante de valores faltantes (MNAR, identificados en semanas anteriores).

Al excluir los nulos con `df[df["correo_cliente"].notna()].duplicated(subset=["correo_cliente"], keep=False).sum()`, el número real de registros con correo duplicado es **26,128**.



In [32]:
duplicados_correo_distinto_pedido = df[
    df["correo_cliente"].notna() & 
    df.duplicated(subset=["correo_cliente"], keep=False)
].sort_values("correo_cliente")

# Filtra solo los casos donde el id_pedido SÍ es distinto (cliente recurrente real)
recurrentes_reales = duplicados_correo_distinto_pedido[
    duplicados_correo_distinto_pedido.duplicated(subset=["correo_cliente", "id_pedido"], keep=False) == False
]

print(recurrentes_reales[["id_pedido", "correo_cliente", "fecha_pedido"]].head(10))

Empty DataFrame
Columns: [id_pedido, correo_cliente, fecha_pedido]
Index: []


### Duplicados por correo encontrados

Al ejecutar `df.duplicated(subset=["correo_cliente"], keep=False).sum()` sin filtrar, se obtuvieron 197,944 registros, cifra inflada porque pandas trata los valores `NaN` como iguales entre sí y `correo_cliente` tiene una alta proporción de valores faltantes (MNAR).

Al excluir los nulos, el número real de registros con correo repetido es **26,128**.

Al revisar estos casos, se confirmó que **el 100% corresponde a duplicados exactos de fila completa** (mismo `id_pedido`, mismo correo, mismos valores en todas las columnas) — no se encontró ni un solo caso de cliente recurrente legítimo (mismo correo con `id_pedido` distinto). Ejemplos verificados:
- `cliente100017@novamarket.com.co` (pedido `P200017`): registrado 2 veces con valores idénticos.
- `cliente100092@yahoo.es` (pedido `P200092`): registrado 2 veces con valores idénticos.

**Conclusión**: los duplicados por correo no representan un fenómeno nuevo, sino una confirmación adicional del hallazgo de duplicados exactos (18,410 filas). Se recomienda tratarlos en conjunto con `df.drop_duplicates()`, priorizando `id_pedido` como criterio, y no requieren una estrategia de deduplicación separada por correo.

### Aclaración: por qué los conteos de duplicados dan números distintos

Al comparar los tres resultados obtenidos, es importante entender que cada uno mide algo diferente:

| Código | Qué compara | Resultado |
|---|---|---|
| `df.duplicated().sum()` | Las 22 columnas deben ser idénticas entre filas | 18,410 |
| `df.duplicated(subset=["correo_cliente"], keep=False).sum()` (sin filtrar) | Solo `correo_cliente`, incluyendo valores NaN | 197,944 |
| Mismo código, excluyendo NaN con `.notna()` | Solo `correo_cliente`, con correos reales | 26,128 |

**Diferencia entre 18,410 y 26,128**: son preguntas distintas. `df.duplicated()` busca filas completas idénticas en todas las columnas, mientras que `duplicated(subset=["correo_cliente"])` busca solo coincidencia en el correo, sin importar el resto. Además, `keep=False` marca **todas** las filas de un grupo duplicado (si un valor aparece 3 veces, marca las 3), mientras que `duplicated()` sin ese parámetro solo marca las copias adicionales, dejando sin marcar la primera aparición. Al verificar los casos uno por uno, se confirmó que en este dataset ambos números describen el mismo fenómeno de fondo: duplicados exactos de fila completa.

**Diferencia entre 197,944 y 26,128**: esta sí es un artefacto técnico, no un hallazgo real. Pandas trata todos los valores `NaN` como iguales entre sí al buscar duplicados. Como `correo_cliente` tiene 171,816 registros con valor faltante (identificados como MNAR en semanas anteriores), pandas marca esas filas como "duplicadas" entre sí simplemente porque todas están vacías — no porque compartan un correo real. Por eso se filtró con `.notna()` antes de comparar, quedando el

## 4. Eliminacion de duplicados

### Regla de conservación de duplicados

Para los 18,410 duplicados exactos detectados, se conservará la **primera aparición de cada fila** (`keep="first"` en `df.drop_duplicates()`), ya que se confirmó que cada grupo duplicado es idéntico en las 22 columnas — incluyendo `fecha_pedido` — por lo que no existe una copia "más reciente" o "más completa" entre ellas; ambas contienen exactamente la misma información.

**Regla secundaria (no aplica en este caso)**: si en una revisión posterior se encontraran duplicados con diferencias en `fecha_pedido` o en la cantidad de valores vacíos, se conservaría la fila con la fecha más reciente y, en caso de empate, la que tenga menos valores nulos (`df.isnull().sum(axis=1)` más bajo). Esta regla queda documentada como respaldo, aunque no fue necesaria para los duplicados exactos de este dataset.

In [34]:
filas_antes = df.shape[0]
print("Filas antes de eliminar duplicados:", filas_antes)

Filas antes de eliminar duplicados: 589386


### Empiesa la eliminacion de duplicados !!!!

In [35]:
df = df.drop_duplicates(keep="first")

In [36]:
filas_despues = df.shape[0]
print(f"Se eliminaron {filas_antes - filas_despues} filas duplicadas.")

Se eliminaron 18410 filas duplicadas.


Se omitió la línea `drop_duplicates(subset=["correo_cliente"], keep="last")` sugerida en la guía, ya que `correo_cliente` tiene una alta proporción de valores faltantes (MNAR) y esta operación trataría todos los NaN como duplicados entre sí, eliminando incorrectamente ~171,815 registros válidos con correo faltante. Se confirmó previamente que los duplicados por correo coinciden al 100% con los duplicados exactos de fila completa, por lo que basta con `drop_duplicates(keep="first")` para cubrir ambos casos sin riesgo de pérdida de datos.

In [37]:
print("Duplicados exactos:", df.duplicated().sum())
print("Duplicados por correo (sin filtrar):", df.duplicated(subset=["correo_cliente"], keep=False).sum())

Duplicados exactos: 0
Duplicados por correo (sin filtrar): 166470


In [39]:
print("Duplicados por correo (excluyendo NaN):", df[df["correo_cliente"].notna()].duplicated(subset=["correo_cliente"], keep=False).sum())

Duplicados por correo (excluyendo NaN): 0


In [40]:
for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]:
    print(col, df[col].nunique())

ciudad_tienda 12
categoria_producto 6
canal_compra 4
metodo_pago 4


### Verificación final de duplicados

Tras aplicar `drop_duplicates(keep="first")`, el dataset quedó sin duplicados exactos (0) y sin duplicados reales por correo (0, excluyendo valores NaN). El conteo de 166,470 en `duplicated(subset=["correo_cliente"], keep=False)` sin filtrar corresponde al efecto esperado de los valores NaN en `correo_cliente` (aún pendientes de tratamiento como parte del análisis MNAR de semanas anteriores), no a duplicados sin resolver.

Las 4 columnas categóricas (`ciudad_tienda`, `categoria_producto`, `canal_compra`, `metodo_pago`) muestran conteos de valores únicos coherentes con la revisión de la lección anterior: 12, 6, 4 y 4 respectivamente.

## 5. Crear el nuevo df

In [41]:
df.to_csv("S06_PD_Diaz_DatasetDepurado.csv", index=False)

In [42]:
df_verificacion = pd.read_csv("S06_PD_Diaz_DatasetDepurado.csv")

print("Duplicados exactos:", df_verificacion.duplicated().sum())
for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]:
    print(col, sorted(df_verificacion[col].dropna().unique()))

Duplicados exactos: 0
ciudad_tienda ['Barranquilla', 'Bogotá', 'Bucaramanga', 'Cali', 'Cartagena', 'Cúcuta', 'Ibagué', 'Manizales', 'Medellín', 'Pereira', 'Santa Marta', 'Villavicencio']
categoria_producto ['Belleza', 'Deportes', 'Electrónica', 'Hogar', 'Juguetería', 'Moda']
canal_compra ['App', 'Marketplace', 'Tienda', 'Web']
metodo_pago ['Efectivo', 'Paypal', 'Tarjeta', 'Transferencia']
